Refactor DLWP wrapper and add batching support

ktangsali opened the pull request:
<!-- markdownlint-disable MD013-->
# Modulus Pull Request

## Description
This PR adds batching support for the DLWP wrapper needed for faster inference and also updates the time stepping such that the time passed to the wrapper is now corresponds to the latest timestep input to the model. 

Closes #137 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->

ktangsali commented:
/blossom-ci

ktangsali commented:
/blossom-ci

dallasfoster commented on modulus/models/fcn_mip_plugin.py:
```diff
@@ -159,21 +159,25 @@ def __init__(
-            input_list[i] = input_list[i].reshape(1, num_chans, 6, 64, 64)
+        M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
+        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
```
I found a bug while using Pytorch 1.13.1 that we should consider:

```python 
input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
```

gives the error

```
NotImplementedError: Could not run 'aten::sparse_dim' with arguments from the 'CUDA' backend
```

while

```python
temp = input_list[i].reshape(num_ens * num_chans, -1)
temp = (M @ temp.T).T
```

does not. This is likely due to some limitation in matrix batch multiplication for sparse tensors in Pytorch, Consider revising here and in `prepare_output`.

ktangsali replied:
Great comment. I added a fix that just reshapes the data to a 2d array and then changes the shape again after multiplication. Can you have a look again?

dallasfoster commented on modulus/models/fcn_mip_plugin.py:
```diff
@@ -159,21 +159,25 @@ def __init__(
+        M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
+        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
+        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
```
Can we abstract the size of the underlying grid?

ktangsali replied:
Great comment, done!

dallasfoster commented on modulus/models/fcn_mip_plugin.py:
```diff
@@ -205,25 +210,20 @@ def prepare_input(self, input, time):
-            output_list[i] = output_list[i].reshape(1, num_chans, 721, 1440)
-        output = torch.stack(output_list, dim=1)
+        output = output.reshape(output.shape[0], 2, output.shape[2], -1) @ M.T
+        output = output.reshape(output.shape[0], 2, output.shape[2], 721, 1440)
```
Again, abstract grid size.

dallasfoster reviewed (commented):
Tested, looks good other than comments below.

ktangsali commented:
> Tested, looks good other than comments below.

Thanks @dallasfoster for the review. I have addressed most of the comments. Can you please have a look again?

ktangsali commented:
/blossom-ci

ktangsali commented:
/blossom-ci

dallasfoster commented on modulus/models/fcn_mip_plugin.py:
```diff
@@ -165,8 +167,10 @@ def prepare_input(self, input, time):
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
+        N, T, C = input.shape[0], input.shape[1], input.shape[2]
+        input = input.reshape(N * T * C, -1) @ M.T
+        S = int((M.shape[0] / 6) ** 0.5)
```
So this is a fine choice but maybe why can't we have the size of the grid be defined in the xarray dataset file `self.input_map_wts` and `self.output_map_wts`?

ktangsali replied:
Yeah, let's retain the current way for now, but once we have some conventions on the map files, we can revisit this.

ktangsali commented on modulus/models/fcn_mip_plugin.py:
```diff
@@ -165,8 +167,10 @@ def prepare_input(self, input, time):
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
+        N, T, C = input.shape[0], input.shape[1], input.shape[2]
+        input = input.reshape(N * T * C, -1) @ M.T
```
Thanks @dallasfoster . There seems to be some weird pytorch mismatch that might be causing this as the CI passes. I will have to look into it deeper, but I agree that your proposed method should be much safer. I added it, please take another look.

ktangsali commented:
/blossom-ci

ktangsali commented:
/blossom-ci
